Add a GradientTexture2D editor plugin

fire-forge opened the pull request:
`master` version of https://github.com/godotengine/godot/pull/59291

Closes https://github.com/godotengine/godot-proposals/issues/3800

Adds an editor plugin for editing the `fill_from` and `fill_to` properties of GradientTexture2D in the inspector.

#### Features:
* Handles for moving the points.
* Grid snap toggle with sliders for adjusting grid size (sliders are hidden when grid snap is disabled).
* Button for swapping `fill_from` and `fill_to`.

![image](https://user-images.githubusercontent.com/67974470/158610211-4f0be8a2-c7e1-43d7-8738-32056809c1b9.png) ![image](https://user-images.githubusercontent.com/67974470/158610282-8fc11462-4853-4635-b5ef-dd4105c87c02.png)

---------

<details>
<summary>The reason this was originally opened as a draft:</summary>
<br>
I marked this as a draft because I can't figure out how to disable texture repeat. I've tried setting the node's `texture_repeat` to `TEXTURE_REPEAT_DISABLED`, but that has no effect. The Sprite2D and TextureRect nodes use `draw_texture_rect` as well but don't have this problem, so there must be something else I'm missing.
This is how it looks with an 8x8 texture (left = inspector, right = Sprite2D node): 

![image](https://user-images.githubusercontent.com/67974470/157598359-69d43a11-6c04-435a-9b8b-e0aa4ffeec62.png) ![image](https://user-images.githubusercontent.com/67974470/157598529-4c78cb32-d48a-4187-a5cb-40192460b86c.png)
</details>

fire-forge committed:
Add GradientTexture2D editor plugin

fire-forge commented:
I fixed the texture filter repeat bug, which was caused by an unrelated texture drawing bug.

After this PR is reviewed I can start working on a 3.x version, since GradientTexture2D was backported to 3.5.

groud commented:
This looks quite good. I would change two things maybe:
- I don't think we need different X and Y values for the grid. A single value for both is easier to change and should cover most of the use cases.
- I believe you also should have a way to change the values manually, in case you need some kind of precise values there. So I think we should keep the Vector2 editors for fill_from and fill_to too.

fire-forge commented:
@groud Thanks! I made the changes you suggested. Also, the manual Vector2 editors for `fill_from` and `fill_to` are still there, I just didn't include them in the screenshots.

![image](https://user-images.githubusercontent.com/67974470/158609642-bb434553-38af-4fff-bb7c-fc7f96f15e0c.png)

groud reviewed (approved):
Looks good to me UX wise.
I did not review the code in depth but nothing wrong stood out as wrong.

> Also, the manual Vector2 editors for fill_from and fill_to are still there, I just didn't include them in the screenshots.

Oh sorry for that then, I was not sure if the plugin replaced the properties or was an addition to the inspector. That's perfect then.

KoBeWi commented on editor/plugins/gradient_texture_2d_editor_plugin.cpp:
```diff
@@ -0,0 +1,283 @@
+
+void GradientTexture2DEditorRect::set_texture(Ref<GradientTexture2D> &p_texture) {
+	texture = p_texture;
+	texture.ptr()->connect("changed", callable_mp((CanvasItem *)this, &CanvasItem::update));
```
What's the purpose of using `ptr()` here?

fire-forge replied:
I didn't understand when `.ptr()` should and shouldn't be used at the time I wrote that code. I was getting errors in some places for not using it, so I added it everywhere to get it working, and then forgot to remove it in some places. I removed all the unnecessary `.ptr()` calls now.

fire-forge commented:
> The handles feel somewhat difficult to click when they are at the edges. Not sure if it's fixable, as the square area simply doesn't receive input when clicking outside.

I added margins equal to half of the handle icon size so that the whole handle area can always be clicked.

> Also I managed to move a handle outside the usable area and then I couldn't click it at all. It happened once and I can't reproduce it anymore :/

Did you set the position in the Vector2 editor to something outside of 0 to 1? The handle positions weren't clamped when set outside of mouse input before. I just fixed this, so the handles should never be able to go outside of the texture area even if the value is outside of the 0 to 1 range.

KoBeWi commented on editor/plugins/gradient_texture_2d_editor_plugin.cpp:
```diff
@@ -0,0 +1,283 @@
+	return Vector2(CLAMP(percent.x, 0, 1), CLAMP(percent.y, 0, 1)) * size;
+}
+
+void GradientTexture2DEditorRect::_update_fill_position() {
+	// Update the texture's fill_from/fill_to property based on mouse input.
+	if (handle != HANDLE_NONE) {
```
```C++
if (handle == HANDLE_NONE) {
    return;
}
```
It would be simpler like this. Less indentation.

KoBeWi commented on editor/plugins/gradient_texture_2d_editor_plugin.cpp:
```diff
@@ -0,0 +1,283 @@
+	add_child(texture_editor_rect);
+
+	set_mouse_filter(MOUSE_FILTER_STOP);
+	_set_snap_enabled(false);
```
```suggestion
	_set_snap_enabled(snap_button->is_pressed());
```
It doesn't matter that much, but it's more consistent with snap size.

KoBeWi commented on editor/plugins/gradient_texture_2d_editor_plugin.h:
```diff
@@ -0,0 +1,112 @@
+	Ref<GradientTexture2D> texture;
+	UndoRedo *undo_redo = nullptr;
+	bool snap_enabled = false;
+	float snap_size;
```
```suggestion
	float snap_size = 0;
```

fire-forge commented:
I updated it with the suggested code changes :)

akien-mga commented on editor/plugins/gradient_texture_2d_editor_plugin.cpp:
```diff
@@ -0,0 +1,285 @@
+Point2 GradientTexture2DEditorRect::_get_handle_position(const Handle p_handle) {
+	// Get the handle's mouse position in pixels relative to offset.
+	const Vector2 percent = p_handle == HANDLE_FILL_FROM ? texture->get_fill_from() : texture->get_fill_to();
+	return Vector2(CLAMP(percent.x, 0, 1), CLAMP(percent.y, 0, 1)) * size;
```
I think you can use `percent.clamp()` directly to do this.

akien-mga commented on editor/plugins/gradient_texture_2d_editor_plugin.cpp:
```diff
@@ -0,0 +1,285 @@
+	Vector2 percent = (get_local_mouse_position() - offset) / size;
+	percent = Vector2(CLAMP(percent.x, 0, 1), CLAMP(percent.y, 0, 1));
+	if (snap_enabled) {
+		percent = Vector2(Math::snapped(percent.x - 0.5, snap_size) + 0.5, Math::snapped(percent.y - 0.5, snap_size) + 0.5);
```
Could also use https://docs.godotengine.org/en/latest/classes/class_vector2.html#class-vector2-method-snapped directly (though it might not be a lot more readable here).

akien-mga commented on editor/plugins/gradient_texture_2d_editor_plugin.cpp:
```diff
@@ -0,0 +1,285 @@
+		case NOTIFICATION_ENTER_TREE:
+		case NOTIFICATION_THEME_CHANGED: {
+			checkerboard->set_texture(get_theme_icon(SNAME("GuiMiniCheckerboard"), SNAME("EditorIcons")));
+		} break;
```
```suggestion
		} break;

```

Separation lind between breaks in big `_notification` switches for readability.

akien-mga commented:
Thanks!
